### PR TITLE
Replace units.yaml with new parameter definition logic

### LINKED ISSUE #416: Remove reliance on Units.ts and UnitKeyName
## Story
As a user I want to be able to create a new parameter without having to edit the core Impact Framework CLI.

## Rationale
We've been discussing the many issues with units.yml, including the core issue that a plugin author should not have to make a PR to the core framework repo to adjust units.yml so they can use their plugin.

At the same time, there is an additional piece of code in the core framework that strongly types parameter names in terms of a TypeScript type. See:
- https://github.com/Green-Software-Foundation/if/blob/dev/src/types/units.ts
- https://github.com/Green-Software-Foundation/if/blob/dev/src/types/model-interface.ts

So regardless of if we complete the work in units.yml, with this strongly typed requirement in the core IF framework a plugin author will still need to submit a pull request to the core library to make their plugin work.

It seems at odds with the work being don't to normalize and standardize units.yml and against the design philosophy of [normalized decentralized development](https://github.com/Green-Software-Foundation/if-docs/blob/restructure/docs/major-concepts/design-philosophy.md#neutrality).

## Implementation details
Remove `units.ts` and make the Params type more generalized and open.

```
export type Params = {
  string: any
};
```

It should not affect any existing code to remove this constraint. Given we have such robust conversation and a process being built around parameters, naming, overriding, and development experience, nothing is gained from having a type that insists on the name being specified in a TS file before it can be used as a params object.

## Priority
5/5 - Required for hackathon participants.

## Scope
Model interface.

## Size
Since we are making the type more generalized and open, this change won't affect the existing plugins. They should all continue working. This is a small task 1d max.

## What does "done" look like?
- [ ] Can create a manifest file with a parameter that does not exist in untis.ts and it won't error out.

## Does this require updates to documentation or other materials??
Yes, to the plugin interface docs.

## What testing is required?
Standard unit tests passing for model plugins and IF will suffice.

## Is this a known/expected update?
Not explicitly discussed, related to these issues:
https://github.com/Green-Software-Foundation/if/issues/412
https://github.com/Green-Software-Foundation/if/issues/403
https://github.com/Green-Software-Foundation/if/issues/403

### PR DESCRIPTION
**Pull request by jmcook1186**
### Types of changes

- [x] Enhancement (project structure, spelling, grammar, formatting)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### A description of the changes proposed in the Pull Request
This PR removes all logic related to `units.ts` and `untis.yml` and instead:

- uses parameters.ts to export canonical parameter definitions
- allows users to add new parameters to the canonical set via the manifest file
- allows users to override the canonical set by providing their own file (.ts, .js or .json)
- provides necessary validation and warning logs
- updates tests and fixes breakages caused by logic change

This does NOT yet pull the necessary file from an npm package - instead it uses a local `param.ts` file in place of the old `units.yml`. The user can append to that file via the manifest file or provide a completely separate one via the CLI.

closes #416 and #403

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -13,6 +13,8 @@
     "url": "https://github.com/Green-Software-Foundation/if/issues/new?assignees=&labels=feedback&projects=&template=feedback.md&title=Feedback+-+"
   },
   "dependencies": {
+    "@grnsft/if-models": "^0.1.3",
```
Comment (narekhovhannisyan) [inline package.json:16]: please remove

File: src/__tests__/unit/util/units-dealer.test.ts [hunk 2]
```diff
@@ -1,25 +1,13 @@
...
-      if (process.env.REJECT_READ === 'true') {
-        return Promise.reject(new Error('mock-error'));
-      }
-
-      return Promise.resolve({
-        'cpu-util': {
-          aggregation: 'avg',
-        },
-      });
+jest.mock('../../../types/units', () => {
```
Comment (narekhovhannisyan) [inline src/__tests__/unit/util/units-dealer.test.ts:3]: please check the entire test file

File: src/config/config.ts [hunk 3]
```diff
@@ -17,6 +17,11 @@ export const CONFIG = {
         description:
           'Path to the output IMPL file where the results as saved, if none is provided it prints to stdout.',
       },
+      overrideParams: {
```
Comment (narekhovhannisyan) [inline src/config/config.ts:20]: please check if it works with 'override-params'

File: src/index.ts [hunk 4]
```diff
@@ -31,8 +31,11 @@ const impactEngine = async () => {
   const processParams = parseProcessArgument();
 
   if (processParams) {
-    const {inputPath, outputPath} = processParams;
+    const {paramPath, inputPath, outputPath} = processParams;
     const rawImpl = await openYamlFileAsObject<Impl>(inputPath);
+    if (!(paramPath === null || paramPath === undefined)) {
+      console.log(OVERRIDE_WARNING);
+    }
```
Comment (narekhovhannisyan) [inline src/index.ts:38]: if (paramPath) {
  console.warn
}

File: src/index.ts [hunk 5]
```diff
@@ -43,14 +46,19 @@ const impactEngine = async () => {
...
+      impl,
+      modelsHandbook,
+      paramPath
+    );
+    await computeInstance.synchronizeParameters();
+    const outputData = await computeInstance.compute();
 
     if (!outputPath) {
-      console.log(JSON.stringify(ompl));
+      console.log(JSON.stringify(outputData));
```
Comment (narekhovhannisyan) [inline src/index.ts:58]: console.log(JSON.stringify(outputData, null, 4))

File: src/lib/supercomputer.ts [hunk 6]
```diff
@@ -1,10 +1,11 @@
 import {ModelsUniverse} from './models-universe';
 import {Observatory} from './observatory';
 import {aggregate} from './aggregator';
-
 import {ERRORS} from '../util/errors';
-
 import {STRINGS} from '../config';
+import {Parameter} from '../types/units';
+import {parameters} from '../config/params';
+const fs = require('fs');
```
Comment (narekhovhannisyan) [inline src/lib/supercomputer.ts:8]: can be moved to the top

Comment (narekhovhannisyan) [reply src/lib/supercomputer.ts:8]: import * as fs from 'node:fs'

File: src/lib/supercomputer.ts [hunk 7]
```diff
@@ -17,6 +18,7 @@ import {
   hasInputs,
 } from '../types/impl';
 import {ModelParams} from '../types/model-interface';
+import {warn} from 'console';
```
Comment (narekhovhannisyan) [inline src/lib/supercomputer.ts:21]: can be used directly

File: src/lib/supercomputer.ts [hunk 8]
```diff
@@ -1,10 +1,11 @@
 import {ModelsUniverse} from './models-universe';
 import {Observatory} from './observatory';
 import {aggregate} from './aggregator';
-
 import {ERRORS} from '../util/errors';
-
 import {STRINGS} from '../config';
```
Comment (narekhovhannisyan) [inline src/lib/supercomputer.ts:5]: please export params from index file, and use here

File: src/util/units-dealer.ts [hunk 9]
```diff
@@ -1,35 +1,16 @@
...
-
   return {
     /**
      * Returns aggregation method for given `unitName`. If doesn't exist then returns value `sum`.
      */
-    askToGiveMethodFor: (unitName: UnitKeyName) => {
-      if (unitsStack[unitName]) {
-        return unitsStack[unitName].aggregation;
+    askToGiveMethodFor: (unitName: string) => {
+      if (Object.prototype.hasOwnProperty.call(parameters, unitName)) {
```
Comment (narekhovhannisyan) [inline src/util/units-dealer.ts:12]: unit name in parameters

File: package.json (hunk 1 above)
Comment (jmcook1186) [reply package.json:16]: done

File: src/lib/supercomputer.ts (hunk 6 above)
Comment (jmcook1186) [reply src/lib/supercomputer.ts:8]: done

File: src/lib/supercomputer.ts (hunk 7 above)
Comment (jmcook1186) [reply src/lib/supercomputer.ts:21]: oop, yep, thanks

File: src/config/config.ts (hunk 3 above)
Comment (jmcook1186) [reply src/config/config.ts]: works, thanks

File: src/index.ts (hunk 5 above)
Comment (jmcook1186) [reply src/index.ts:71]: done

**Review by narekhovhannisyan — APPROVED** [review]

**Pull request merged**